contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -1497,12 +1501,73 @@ pre_transform_target_entry(ResTarget *res, ParseState *pstate,
 			 * sourcetext
 			 */
 			if (list_length(cref->fields) == 1 &&
-				IsA(linitial(cref->fields), String))
+				IsA(linitial(cref->fields), String)) 
 			{
 				identifier_name = strVal(linitial(cref->fields));
 				alias_len = strlen(identifier_name);
 				colname_start = pstate->p_sourcetext + res->location;
 			}
+			/*
+			 * This condition will preserve the case of column name when there are more than
+			 * one cref->feilds. For instance, Queries like 
+			 * select [database].[schema].[table].[column] from table.
+			 * select [schema].[table].[column] from table.
+			 * selec [t].[column] from table as t
+			 * Case 1: Handle the cases when column name is passed with no delimiters
+			 * For example, select ABC from table
+			 * Case 2: Handle the cases when column name is delimited with sq and dq.
+			 * In such cases, we are checking if no. of dq or sq are even or not. When sq are odd,
+			 * we are not tracing number of sqb and dq within sq, similar with the case of dq.
+			 * For instance, Queries like select "AF bjs'vs] " from table.
+			 * Case 3: Handle the case when column name is delimited with sqb. When number of sqb
+			 * are zero, it means we are out of sqb.
+			 */
+			if(list_length(cref->fields) > 1 &&
+				IsA(llast(cref->fields), String))
+			{
+				identifier_name = strVal(llast(cref->fields));
+				alias_len = strlen(identifier_name);
+				colname_start = pstate->p_sourcetext + res->location;
+				while(true)
+				{	
+					if(open_square_bracket == 0 && single_quotes % 2 == 0 && *colname_start == '"')
+					{
+						double_quotes++;
+					}
+					if(open_square_bracket == 0 && double_quotes % 2 == 0 && *colname_start == '\'')
+					{
+						single_quotes++;
+					}
+					/* To check how many open sqb are present in sourcetext. */
+					if(double_quotes % 2 == 0 && single_quotes % 2 == 0 && *colname_start == '[')
+					{
+						open_square_bracket++;
+					}
+					if(double_quotes % 2 == 0 && single_quotes % 2 == 0 && *colname_start == ']')
+					{
+						open_square_bracket--;
+					}
+					/*
+					 * last_dot pointer is to trace the last dot in the sourcetext,
+					 * as last dot indicates the starting of column name.
+					 */
+					if(open_square_bracket == 0 && double_quotes % 2 == 0 && single_quotes % 2 == 0 && *colname_start == '.')
+					{
+						last_dot = colname_start;
+					}
+					/* 
+					 * If there is no open sqb, there are even no. of sq or dq and colname_start is at
+					 * space or comma, it means colname_start is at the end of column name.
```

Comment:
To protect from unexpected failure, lets try to find the length of each component and assert the length that it should be <64